### PR TITLE
feat(dap): switch to using `lldb` adapter, dap-client configuration from config, load `launch.json` by default, `probe-rs` support

### PR DESCRIPTION
**Pull request by Andrew-Collins**
###### Description of changes
Added support for the `probe-rs` adapter, and in doing so implemented (at least in spirit) #29 and #66.

There are a lot of possible implementations, so I'll add some comments to the diffs discussing what I came up with.
###### Things done
- Switch to using `lldb` adapter instead of `rt_lldb`.
- Added `dap.configuration` entry to config, allowing for configuration of the dap-client.
  - Default for this entry loads `launch.json` with the fallback of a basic config of the `lldb` adapter.
- Support `probe-rs` adapter when configuring the dap-client. 

- [x] Tested, as applicable:
  - [x] Manually
  - [ ] Added `busted` specs
- [x] Updated [CHANGELOG.md](https://github.com/mrcjkb/rustaceanvim/blob/master/CHANGELOG.md) (if applicable).
- [x] Fits [CONTRIBUTING.md](https://github.com/mrcjkb/rustaceanvim/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: doc/rustaceanvim.txt
Comment (Andrew-Collins) [inline doc/rustaceanvim.txt:1]: These are just basic fields, directly used in dap.lua.

A comprehensive list could be constructed for lldb and probe-rs adapters, but I chose not to do this and it would be a fair amount of work to implement and maintain, for not much gain imo.

File: lua/rustaceanvim/config/internal.lua
Comment (Andrew-Collins) [inline lua/rustaceanvim/config/internal.lua:289]: I chose to load `launch.json` in the `configuration` field instead of in `dap.lua` because imo it is more transparent, and gives the user full control.

This can be easily moved if required.

File: lua/rustaceanvim/dap.lua
Comment (Andrew-Collins) [inline lua/rustaceanvim/dap.lua:269]: This can be moved into the `dap.adapter` configuration for better transparency/control.

It won't do any harm being here, but it will make things 'just work' so it could cause some confusion for users who are setting up their own `lldb` adapter.

Configuration of the `lldb` adapter happens here, instead of in the `dap.adapter` configuration function as an `lldb` adapter can be added at any point in the users configuration.

However,

**Comment by mrcjkb** [comment]
Wow, this looks great, thanks! 😄 

I'll try to find some time to review this in detail soon.

**Comment by Andrew-Collins** [comment]
No worries. :smile:  

Feel free to ask for things to be moved around; I did things in a way that made sense to me, but happy to defer to your opinion/vision.

**Comment by mrcjkb** [comment]
cc @richchurcher tagging you because you were interested in this, too 😄

File: lua/rustaceanvim/config/init.lua
Comment (mrcjkb) [inline lua/rustaceanvim/config/init.lua:129]: question: Is this `string` or `string[]`?

File: lua/rustaceanvim/config/internal.lua
Comment (mrcjkb) [inline lua/rustaceanvim/config/internal.lua:276]:
```suggestion
      -- default
      ---@type DapClientConfig
      local dap_config = {
```

File: lua/rustaceanvim/config/internal.lua
Comment (mrcjkb) [inline lua/rustaceanvim/config/internal.lua:291]:
```suggestion
        if pre_launch[k] == nil or not vim.deep_equal(pre_launch[k], v) then
```

File: lua/rustaceanvim/dap.lua
Comment (mrcjkb) [reply lua/rustaceanvim/dap.lua:269]: It looks like some of your text got cut off?

> Configuration of the lldb adapter happens here, instead of in the dap.adapter configuration function as an lldb adapter can be added at any point in the users configuration.

To solve this, we can call `types.evalueate(config.dap.adapter)` within the `dap.start()` function, and pass it down to wherever it's needed as a function parameter.

I can implement that after this is merged though :smile:

File: lua/rustaceanvim/config/init.lua
Comment (Andrew-Collins) [reply lua/rustaceanvim/config/init.lua:124]: Yes, according to `codelldb` anyway:
https://github.com/vadimcn/codelldb/blob/master/MANUAL.md#custom-launch

File: lua/rustaceanvim/config/init.lua
Comment (Andrew-Collins) [reply lua/rustaceanvim/config/init.lua:130]: I don't use it, lets remove it.

File: lua/rustaceanvim/config/init.lua
Comment (Andrew-Collins) [reply lua/rustaceanvim/config/init.lua:131]: The `probe-rs` site has a bunch of exmaples:
https://probe.rs/docs/tools/debugger/#start-a-debug-session-with-minimum-configuration
Otherwise here is a basic config I am currently using:
https://gist.github.com/Andrew-Collins/1f8ce17ec1372e9f5a24755a3eed89a5

File: lua/rustaceanvim/dap.lua
Comment (Andrew-Collins) [reply lua/rustaceanvim/dap.lua:269]: Whoops, yes it did.

The part that got cutoff was the bit explaining why I don't think that matters.
If the user config comes along later, it will overwrite what `rustacean` has done; which is the intended outcome.

Either way suits me :smile:

File: lua/rustaceanvim/config/init.lua
Comment (Andrew-Collins) [reply lua/rustaceanvim/config/init.lua:129]: It's a string[]

https://github.com/vadimcn/codelldb/blob/master/MANUAL.md#starting-a-new-debug-session

Comment (Andrew-Collins) [reply lua/rustaceanvim/config/init.lua:129]:
```suggestion
---@field initCommands? string[] Initial commands to run, `lldb` clients only
```

File: lua/rustaceanvim/config/init.lua
Comment (Andrew-Collins) [reply lua/rustaceanvim/config/init.lua:130]:
```suggestion
```

File: lua/rustaceanvim/config/init.lua
Comment (Andrew-Collins) [reply lua/rustaceanvim/config/init.lua:131]: I can easily create an example project for you if you like, but that would require having compatible hardware. So let me know what you have access to.

File: lua/rustaceanvim/config/init.lua
Comment (Andrew-Collins) [reply lua/rustaceanvim/config/init.lua]: @mrcjkb this has caused an `undefined-field` warning in `check.lua` where it is validated.
Do you have a preferred way to resolve this? Or can I just get the linter to ignore the line?

Comment (mrcjkb) [reply lua/rustaceanvim/config/init.lua]: Now that it's no longer documented, we can remove it from the validation.

File: lua/rustaceanvim/config/init.lua
Comment (mrcjkb) [reply lua/rustaceanvim/config/init.lua]: No need. The `probe-rs` examples should be enough.
I am hoping to add integration tests for all of the dap features eventually. But if it's dependent on hardware, that might not be a good idea for `probe-rs`.

File: lua/rustaceanvim/config/check.lua
Comment (mrcjkb) [inline lua/rustaceanvim/config/check.lua:126]:
```suggestion
```

**Review by mrcjkb — APPROVED** [review]

**Pull request merged**
